### DISCUSSION TITLE
Manifest v3

### DISCUSSION THREAD

**Discussion by avi12** [crxjs/chrome-extension-tools#79] [General]
### Is your feature request related to a problem? Please describe.

It's not a problem yet, but it will become one in the future.
Manifest v2 will become deprecated at some point, which will be replaced by [Manifest v3](https://developer.chrome.com/docs/extensions/mv3/).
I watched a little bit of the Chrome Dev Summit, and there I heard that the first version of Manifest v3 will hit Chrome Canary (I believe) around mid-January 2021.

### Describe the solution you'd like

Begin reworking the plugin on a separated branch and make it to be the master branch when Manifest v2 will go deprecated.
FYI, Google has made a [migration guide](https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/).

**Comment by rick-stevens** [comment]
V3 has been out for a while now and [Google urges V2 users to migrate](https://developer.chrome.com/docs/extensions/mv2/). Any update on this?

**Comment by jacksteamdev** [comment]
This plugin relies on multi-file inputs in the Rollup config, and the only output format appropriate to Chrome Extensions [that Rollup supports](https://github.com/rollup/rollup/issues/2756) with multi-file inputs is ESM (with mandatory file-splitting). 

This has been a blocking issue for MV3 support, since until now Chrome didn't support [ES Modules for service workers](https://chromestatus.com/feature/4609574738853888). This feature landed in Chrome 91, so we can move ahead with MV3!

I've set aside some time to work on this in August, and updates will be posted here.

**Comment by jacksteamdev** [comment]
MV3 is coming along nicely. You can track our progress on #94

**Comment by jacksteamdev** [comment]
MV3 support is now available in the `main` branch! :tada: We haven't written the docs yet, but no special config is required. Even the simple reloader works, but only if a content script is open.

You can try it out on NPM:

```sh
npm i rollup-plugin-chrome-extension@latest -D
```

All that is required is a valid MV3 `manifest.json` file. If you want to use our JSON schema in VSCode to get auto-completion and validation, add the following to `settings.json`:

```json
{
  "json.schemas": [
    {
      "fileMatch": ["manifest.json"],
      "url": "https://raw.githubusercontent.com/extend-chrome/rollup-plugin-chrome-extension/main/manifest.schema.json"
    }
  ],
}
```

**Reply by milahu** [reply]
schema url is gone, this works

https://raw.githubusercontent.com/extend-chrome/manifest-json-schema/main/schema/manifest.schema.json

... but manifest.js is more flexible

**Comment by avi12** [comment]
How can I have the same code but output to manifest v3 *and* manifest v2, considering that not all browsers support/will support manifest v3?

**Reply by jacksteamdev** [reply]
The manifest is pretty different for MV2 and MV3, but if you can code around the differences, I think it could be done.

It is now possible to write your manifest in JavaScript or TypeScript, so that is where I would start. 

It would look something like...

```typescript
// src/manifest.ts
import { ManifestV3, ManifestV2 } from 'rollup-plugin-chrome-extension'

const mv2: ManifestV2 = { ... }
const mv3: ManifestV3 = { ... }

const manifest = process.env.mv3 ? mv3 : mv2

export default manifest
```

```javascript
// rollup.config.js
// ... do your imports as normal
export default {
  input: 'manifest.ts', // or ['manifest.ts'] or { manifest: `manifest.ts` }
  plugins: [chromeExtension(), simpleReloader(), ...rest]
}
```

YMMV with the actual Chrome Extension code, but please let us know what you come up with!

**Comment by lalwanikamaldev** [comment]
Can we run this command for Angular based (contentscript) chrome extension also ?

npm i rollup-plugin-chrome-extension@latest -D

**Reply by Toumash** [reply]
This is not the best place to ask this, but https://github.com/wxt-dev/wxt/discussions/818
if you have a vite plugin for angular then should be fine i guess
https://www.npmjs.com/package/@analogjs/vite-plugin-angular
